src/verifier/step4/SumcheckLogic.sol

```diff
@@ -0,0 +1,488 @@
+        }
+    }
+
+    function decompress(CompressedUniPoly calldata poly, uint256 hint) public pure returns (UniPoly memory) {
+        // uint256 linear_term = hint - poly.coeffs_except_linear_term[0] - poly.coeffs_except_linear_term[0];
+        uint256 linear_term = addmod(
+            hint,
+            Vesta.negateScalar(
+                addmod(poly.coeffs_except_linear_term[0], poly.coeffs_except_linear_term[0], Vesta.R_MOD)
+            ),
+            Vesta.R_MOD
+        );
+        for (uint256 i = 1; i < poly.coeffs_except_linear_term.length; i++) {
+            // linear_term -= poly.coeffs_except_linear_term[i];
+            linear_term = addmod(linear_term, Vesta.negateScalar(poly.coeffs_except_linear_term[i]), Vesta.R_MOD);
+        }
+
+        uint256[] memory coeffs = new uint256[](poly.coeffs_except_linear_term.length + 1);
+        coeffs[0] = poly.coeffs_except_linear_term[0];
+        coeffs[1] = linear_term;
+
+        for (uint256 i = 1; i < poly.coeffs_except_linear_term.length; i++) {
+            coeffs[i + 1] = poly.coeffs_except_linear_term[i];
+        }
+
+        return UniPoly(coeffs);
+    }
+
+    function toUInt8Array(uint256 input) private pure returns (uint8[] memory) {
+        uint8[] memory result = new uint8[](32);
+
+        bytes32 input_bytes = bytes32(input);
+
+        for (uint256 i = 0; i < 32; i++) {
+            result[i] = uint8(input_bytes[31 - i]);
+        }
+        return result;
+    }
+
+    function toTranscriptBytes(UniPoly memory poly) public pure returns (uint8[] memory) {
+        uint8[] memory result = new uint8[](32 * (poly.coeffs.length - 1));
+
+        uint256 offset;
+        uint8[] memory coeff_bytes = toUInt8Array(poly.coeffs[0]);
+        for (uint256 i = 0; i < 32; i++) {
+            result[i] = coeff_bytes[i];
+        }
+        offset += 32;
+
+        for (uint256 i = 2; i < poly.coeffs.length; i++) {
+            coeff_bytes = toUInt8Array(poly.coeffs[i]);
+            for (uint256 j = 0; j < 32; j++) {
+                result[offset + j] = coeff_bytes[j];
+            }
+            offset += 32;
+        }
+
+        return result;
+    }
+}
+
+library PrimarySumcheck {
+    struct SumcheckProof {
+        PallasPolyLib.CompressedUniPoly[] compressed_polys;
+    }
+
+    function verify(
+        SumcheckProof calldata proof,
+        uint256 claim,
+        uint256 num_rounds,
+        uint256 degree_bound,
+        KeccakTranscriptLib.KeccakTranscript memory transcript
+    ) public pure returns (uint256, uint256[] memory, KeccakTranscriptLib.KeccakTranscript memory) {
+        uint256 e = claim;
+        uint256[] memory r = new uint256[](num_rounds);
+
+        uint8[] memory p_label = new uint8[](1);
+        uint8[] memory c_label = new uint8[](1);
+
+        p_label[0] = 112;
+        c_label[0] = 99;
+
+        require(proof.compressed_polys.length == num_rounds, "Wrong number of polynomials");
```

Comment:
It is very helpful while debugging if we put library and function name into exception string, like following:
`"[PrimarySumcheck::verify] Wrong number of polynomials"`